### ISSUE TITLE
Remote write doesn't handle SigV4 expiration gracefully

### ISSUE THREAD

**Issue by harry671003** [prometheus/prometheus#15186]
### What did you do?

Prometheus logs shows, that when the SigV4 token expires Prometheus doesn't gracefully handle this. The samples are dropped and not retried.

The meric `rate(prometheus_remote_storage_samples_failed_total[5m])` also shows an increase at the same time which indicates the samples were not sent correctly.
<img width="1875" alt="image" src="https://github.com/user-attachments/assets/1435a209-1207-4874-91c7-2395f134da03">

Logs are attached.

### What did you expect to see?

Prometheus should handle token expiry gracefully and shouldn't loose samples.

### What did you see instead? Under which circumstances?

N/A

### System information

Linux

### Prometheus version

```text
Prometheus/2.50.1
```

### Logs

```text
ts=2024-10-17T19:40:40.208Z caller=dedupe.go:112 component=remote level=error remote_name=e94106 url=https://aps-workspaces.<region>.amazonaws.com/<url> msg="non-recoverable error while sending metadata" count=914 err="server returned HTTP status 403 Forbidden: {\"message\":\"The security token included in the request is expired\"}"
```
```

**Comment by machine424** [comment]
Yes, that needs to be gracefully handled, same for the other auth mechanisms.
We'll need a test reproducing this, maybe in https://github.com/prometheus/common/tree/main/sigv4?

**Comment by machine424** [comment]
> [@machine424](https://github.com/machine424) I've done some initial analysis and I'm working on a fix.
> 
> The path you mentioned (github.com/prometheus/common/tree/main/sigv4) doesn't seem to exist. It looks like `storage/remote/client_test.go` is the logical place to add a test for the token-expiration scenario. Does that approach make sense?

My bad, the package lives here now https://github.com/prometheus/sigv4

Let's start with a test+fix in there and then see if we need an extra test in `/prometheus` as well. Do you think this only requires changes on `/prometheus` side?
Thanks.

**Comment by rapphil** [comment]
this PR is supposed to address this issue: https://github.com/prometheus/sigv4/pull/24

**Comment by machine424** [comment]
> this PR is supposed to address this issue: [prometheus/sigv4#24](https://github.com/prometheus/sigv4/pull/24)

Thanks, do you mean the update to v2 is enough and https://github.com/prometheus/sigv4/pull/18 isn't needed?
If yes, It'd still be useful to have a regression test on sigv4 side.

**Comment by bboreham** [comment]
Hello from the bug-scrub!  

https://github.com/prometheus/sigv4/pull/24 was merged as https://github.com/prometheus/sigv4/pull/31, which went into [v0.2.0](https://github.com/prometheus/sigv4/releases/tag/v0.2.0), which went into [Prometheus 3.5.0](https://github.com/prometheus/prometheus/releases/tag/v3.5.0).

@harry671003 can you try v3.5 please?

@rapphil do you think this will fix the problem?

**Comment by bboreham** [comment]
No response for 8 months; closing.
